devlib/instrument/gem5stats.py

```diff
@@ -0,0 +1,152 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import division
+import csv
+import re
+
+from devlib.instrument import Instrument, CONTINUOUS, MeasurementsCsv
+from devlib.exception import TargetError, HostError
+
+class Gem5StatsInstrument(Instrument):
+    '''
+    Instrument enabling power monitoring in Gem5
+
+    This instrument parses the stats file produced by Gem5 and extracts power
+    values. More information on the "stats.txt" file structure can be found 
+    here:
+    
+    http://pages.cs.wisc.edu/~david/courses/cs752/Spring2015/gem5-tutorial/part1/gem5_stats.html
+    
+    This instrument also implements Gem5's ROIs (Regions of Interest). A ROI is
+    an annotation on a time interval marked by start and end cursors. This enables
+    power analysis on a specific time window which spans over the execution of
+    a workload for example. Up to 8 ROIs can be used, with or without 
+    overlapping.
+    '''
+
+    mode = CONTINUOUS
+    dump_head='---------- Begin Simulation Statistics ----------'
+    dump_tail='---------- End Simulation Statistics   ----------'
+    
+    def __init__(self, target, channel_map, stats_file_path):
+        super(Gem5StatsInstrument, self).__init__(target)
+        
+        # Duration since last dump is indexed by 'sim_seconds' in stats.txt
+        self.add_channel('Duration', 'time', 'sim_seconds')
+        # Track power only for now
+        for f in channel_map:
+            self.add_channel(f, 'power', channel_map[f])
+        
+        self._stats_file_path = stats_file_path
+        self._roi_state = [False] * 8
+        self.dumps = []
+
+    def start(self, roi=0):
+        if roi > 7 or roi < 0:
+            raise IndexError('Incorrect ROI number')
+        if self._roi_state[roi]:
+            raise TargetError('ROI {} already running'.format(roi))
+        self._roi_state[roi] = True
+        self.target.execute('m5 roistart {}'.format(roi))
+        self.target.execute('m5 dumpresetstats')
+
+    def stop(self, roi=0):
+        if roi > 7 or roi < 0:
+            raise IndexError('Incorrect ROI number')
+        if not self._roi_state[roi]:
+            raise TargetError('ROI {} not running'.format(roi))
+        self._roi_state[roi] = False
+        self.target.execute('m5 dumpresetstats')
+        self.target.execute('m5 roiend {}'.format(roi))
+
+    def get_data(self, outfile, roi=0):
+        if roi > 7 or roi < 0:
+            raise RuntimeError('Incorrect ROI number')
+        self.target.execute('m5 dumpresetstats')
+        with open(self._stats_file_path, 'r') as f:
+            self._parse_stats(f)
```

Comment:
The way that this is implemented, you could potentially dump for ROIs where were not stopped, yet. Is this something we want? Would it not make more sense check that the ROI is stopped (you already track this in `self._roi_state`)?

Have you considered the memory requirements for this? I worry that loading the data for 10,000 stats dumps into memory will kill many lower spec'd systems.